kroxylicious-filters/kroxylicious-sasl-inspection/src/main/java/io/kroxylicious/filters/sasl/inspection/SaslInspectionFilter.java

```diff
@@ -0,0 +1,361 @@
+        }
+    }
+
+    @NonNull
+    private CompletionStage<ResponseFilterResult> processSuccessfulAuthenticateResponse(ResponseHeaderData header, SaslAuthenticateResponseData response,
+                                                                                        FilterContext context) {
+        if (this.chosenMechanism.isFinished()) {
+            if (this.authorizationIdFromClient == null) {
+                return closeConnectionWithResponse(header, response.setErrorCode(Errors.ILLEGAL_SASL_STATE.code()), context);
+            }
+            LOGGER.atInfo()
+                    .setMessage("Server accepts SASL credentials for client on channel {}, announcing that client has authorizationId {}")
+                    .addArgument(context::channelDescriptor)
+                    .addArgument(() -> this.authorizationIdFromClient)
+                    .log();
+            context.clientSaslAuthenticationSuccess(chosenMechanism.mechanismName(), this.authorizationIdFromClient);
+            resetState(this.clientSupportsReauthentication ? State.ALLOWING_HANDSHAKE_REQUEST : State.DISALLOWING_AUTHENTICATE_REQUEST);
+        }
+        else {
+            this.currentState = State.REQUIRING_AUTHENTICATE_REQUEST;
+        }
+        return context.forwardResponse(header, response);
+    }
+
+    @NonNull
+    private CompletionStage<ResponseFilterResult> processFailedAuthentication(ResponseHeaderData header, SaslAuthenticateResponseData response,
+                                                                              FilterContext context) {
+        Errors error = Errors.forCode(response.errorCode());
+        LOGGER.atInfo()
+                .setMessage("Server rejects SASL credentials with error {} for client on channel {}")
+                .addArgument(error::name)
+                .addArgument(context::channelDescriptor)
+                .log();
+        context.clientSaslAuthenticationFailure(chosenMechanism.mechanismName(), this.authorizationIdFromClient, error.exception());
+        resetState(State.REQUIRING_HANDSHAKE_REQUEST);
+        return context.responseFilterResultBuilder()
+                .forward(header, response)
+                .withCloseConnection()
+                .completed();
+    }
+
+    @NonNull
+    private static CompletionStage<RequestFilterResult> closeConnectionWithShortCircuitResponse(FilterContext context, ApiMessage response) {
+        return context.requestFilterResultBuilder()
+                .shortCircuitResponse(response)
+                .withCloseConnection()
+                .completed();
+    }
+
+    @NonNull
+    private CompletionStage<ResponseFilterResult> closeConnectionWithResponse(ResponseHeaderData header, ApiMessage response, FilterContext context) {
+        LOGGER.error("Unexpected {} response while in state {}. This should not be possible. Closing connection.",
```

Comment:
running with coverage in IDE showed that these unexpected state transition cases aren't covered, in other state machines we sometimes have test-only mechanisms to force it into a particular state so we can exercise these bits of logic.